### ISSUE TITLE
Would you mind wrapping up a 1.4.3 release?

### ISSUE THREAD

**Issue by odrotbohm** [JavaMoney/jsr354-ri#383]
There hasn't been a release in 2,5 years and the branch for 1.4.3 contains quite a few much awaited, low-level fixes (see below), some of them almost 3 years old. Do you think it'd make sense to wrap up the release to ship these?

* #361 
* #357

**Comment by keilw** [comment]
Unfortunately there was not enough interest to aim for a 2.x release anytime soon.
Without #370 it does not make sense to release 1.4.3, most of the others in https://github.com/JavaMoney/jsr354-ri/milestones/1.4.3 could probably be postponed.

**Comment by otaviojava** [comment]
@keilw 
But why not release what we have and then release a version with the others fixes?

**Comment by ratoaq2** [comment]
It would be good to have a 1.4.3 release when that blocker issue is completed.

About a 2.x release, because of this major jakarta namespace change, from my POV, it would be better that 2.0 is exactly the same as 1.4.3, with only the javax -> jakarta imports.

That makes upgrade paths easier.

**Comment by keilw** [comment]
The jakarta namespace is not necessarily up to a 2.x release because that requires a whole new JSR with a new number. Which so far there has not been enough interest and echo by JCP members and while @otaviojava and I are good to help as Maintenance Lead, we would not file a new JSR without a solid and committed EG. 
After the takeover by Accenture it looks like Trivadis has not renewed its JCP Membership: https://jcp.org/en/participation/members/T Nor has parent company Accenture although it used to be a JCP member supporting e.g. the [Portlet Spec](https://jcp.org/en/jsr/proposalDetails?id=168).

If either of you in this thread knows a company that would be happy and able to join as co-spec lead, then we could think of a new JSR, otherwise without at least one Corporate member involved, I guess it's maintenance only.

**Comment by ratoaq2** [comment]
What I meant was to have a release that doesn't depend on old `javax.annotation`.

Not actually moving javax money to jakarta money

**Comment by keilw** [comment]
There will never be a jakarta money, the JSR keeps javax.

#381 would involve using Jakarta EE 9, 

The 4 failed tests must pass, please propose PRs if you find a solution, some may even have something to do with changes to reflection in later Java versions.

**Comment by ratoaq2** [comment]
The 4 failing tests seems related to ServiceLoader not finding the `org.javamoney.moneta.function.TestRoundingProvider` configured in `META-INF/services/javax.money.spi.RoundingProviderSpi` from the tests.

I believe the issue might be related to how Java 9 module system changed the ServiceLoader. I'm still not familiar with java modules but if I find a solution I'll propose a PR for them

**Comment by ratoaq2** [comment]
Just got my local env to quickly look at it
Related to the failing tests in `MonetaryRoundingsTest`:

On java 17, the `java.util.ServiceLoader` is indeed ignoring `TestRoundingProvider` because it is in a named module ( `(java.lang.Module) module org.javamoney.moneta`)

![image](https://user-images.githubusercontent.com/9292886/213300285-0bc4293c-d3d3-4d85-926c-a74ab88d3eb9.png)
![image](https://user-images.githubusercontent.com/9292886/213300332-f6bab7d6-e03f-42a9-acef-0a091730503c.png)

**Comment by keilw** [comment]
And that's not the only place. 

loading resources like the IMF defaults also fails for JPMS related reasons, but that resolved (by putting the files into the package structure (similar to suggestions in https://stackoverflow.com/questions/61531317/how-do-i-determine-the-correct-path-for-fxml-files-css-files-images-and-other/61531318#61531318) still won't get past #353. 
The timeout persists

**Comment by keilw** [comment]
I'

> Just got my local env to quickly look at it Related to the failing tests in `MonetaryRoundingsTest`:
> 
> On java 17, the `java.util.ServiceLoader` is indeed ignoring `TestRoundingProvider` because it is in a named module ( `(java.lang.Module) module org.javamoney.moneta`)
> 
> ![image](https://user-images.githubusercontent.com/9292886/213300285-0bc4293c-d3d3-4d85-926c-a74ab88d3eb9.png) ![image](https://user-images.githubusercontent.com/9292886/213300332-f6bab7d6-e03f-42a9-acef-0a091730503c.png)

@ratoaq2 Thanks for the research. Any remedy for that?

**Comment by keilw** [comment]
Guys, 
This is more or less an "epic" if you want, but I also added it to the [1.4.3](https://github.com/JavaMoney/jsr354-ri/milestones/1.4.3) milestone issues.

Especially the two with "help wanted" can use some help, as both the proper configuration and hopefully a solution to fetching both ECB and IMF exchange rates at runtime, are the most critical issues, while others like Indian formatting mostly concern fixing unit tests and that seems very advanced.

Once those are fixed, we can release 1.4.3.

**Comment by otaviojava** [comment]
IMHO: I would put both services out. 
We don't provide those services, and we don't have any control, such as retrocompability.

**Comment by keilw** [comment]
@otaviojava No we shouldn't drop IMF either, but properly use the REST service, especially with all the Jakarta/Enterprise people involved ;-)
https://datahelp.imf.org/knowledgebase/articles/1968408-how-to-use-the-api-python-and-r
Let's use the API in Java instead of the old download option.

**Comment by keilw** [comment]
With #357 addressed (still waiting for your review) there are at least 2 major showstoppers that need to be addressed:

- #353 
- #374 (affecting at least ECB-HIST, the other two providers look OK in the examples)
For experimental purposes I deactivated IMF until #353 can be solved. The ECB providers for current and 90 days look fine, while ECB-HIST fails for any given date, even though a huge XML is present from around 1999 till now.

**Comment by kewne** [comment]
@keilw I've been looking into the test failures and it seems that, again, the issue is the JPMS...
Some of the failing tests rely on a provider for ServiceLoader that is defined in the test sources, which is not loaded.

From what I've found, the tests apparently run in the main sources' module, which does not declare this test provider.
While there are some solutions floating around that involve patching the module, IDEs don't seem to support them, which I think makes them unsustainable.

What I think *could* work is moving the tests that rely on service loaders to a separate source tree, with a separate module definition, which I've tested and works.
Let me know if this is acceptable and I'll submit a PR for it.

**Comment by keilw** [comment]
@kewne If you can, please provide the PR.

**Comment by keilw** [comment]
Closing as the release is about to be published.

**Comment by keilw** [comment]
Just in case @odrotbohm the 1.4.3 release was published 2 days ago after all the critical issues, especially around exchange sources could be resolved.

**Comment by keilw** [comment]
@odrotbohm We don't plan to use an odd/even pattern like the Linux Kernel, but please use release [1.4.4 ](https://github.com/JavaMoney/jsr354-ri/releases/tag/1.4.4) because in 1.4.3 the IMF exchange rate provider was still disabled to avoid the prior timeouts.

I assume not too many people started using 1.4.3 yet ;-)

**Comment by saw303** [comment]
> I assume not too many people started using 1.4.3 yet ;-)

@keilw well, in times of dependabot & co. your assumption might be wrong. Anyway, thank you for that information.

**Comment by keilw** [comment]
@saw303 is it possible to add upgrade recommendations for libraries like Moneta to Dependabot?

**Comment by saw303** [comment]
@keilw Yes, I guess it is. At least according to the docs https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuring-dependabot-version-updates#checking-the-status-of-version-updates
